model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/liquibase/connection/DefaultLiquibaseConnectionProvider.java

```diff
@@ -91,18 +96,46 @@ private void exitLiquibaseScope(String scopeId) {
         } catch (Exception e) {
             throw new RuntimeException("Failed to exist scope: " + e.getMessage(), e);
         }
+        // To avoid cached instances from the just closed scope.
+        // Must be called only after exiting the scope.
+        SqlGeneratorFactory.reset();
     }
 
     private String enterLiquibaseScope() {
         String scopeId;
         final Map<String, Object> scopeValues = new HashMap<>();
         // Setting the LoggerUIService here prevents Liquibase from logging each change set to the console using java.util.Logging in the Quarkus setup
         scopeValues.put(Scope.Attr.ui.name(), new LoggerUIService());
+
+        // This uses the same serviceloader as the parent (that is replaced in Keycloak Quarkus)
+        // and prevents any class from the legacy JPA store to leak into the service discovery.
+        // This can be removed once the legacy JPA provider is no longer around in both the
+        // Wildfly and Quarkus distributions of Keycloak.
+        ServiceLocator serviceLocator = Scope.getCurrentScope().getServiceLocator();
+        scopeValues.put(Scope.Attr.serviceLocator.name(), new ServiceLocator() {
+            @Override
+            public int getPriority() {
+                return serviceLocator.getPriority();
+            }
+
+            @Override
+            public <T> List<T> findInstances(Class<T> interfaceType) throws ServiceNotFoundException {
+                List<T> instances = serviceLocator.findInstances(interfaceType);
+                // prevent any class from the legacy JPA store to leak into the service discovery
+                instances = instances.stream().filter(t -> !t.getClass().getPackage().getName().startsWith("org.keycloak.connections.jpa")).collect(Collectors.toList());
```

Comment:
Please use a constant for the package name.

Also a nit:
```suggestion
                instances = instances.stream().filter(t -> !t.getClass().getPackageName().startsWith("org.keycloak.connections.jpa")).collect(Collectors.toList());
```

Reply:
The `getPackageName()` is only available as of JDK9, therefore not used here (and Maven/IDE will complain), as code aligns with JDK8.

Constant will be added.